Store locations of longidents components

voodoos opened the pull request:
This PR adds location information to individual longident's components. This information is used extensively by Merlin (for example to identify the part of the path on which the cursor is placed) and other tools, but is impossible to reconstruct without re-parsing the identifiers. Having such locations would allow the removal of several heuristics and custom re-parsing in Merlin. More importantly, it is required if we plan to index modules appearing in longidents so that they can be searched project-wide (and renamed). It should also make the life of other tools manipulating the syntax easier (cc @Julow for ocamlformat).

I opened this PR as a draft because there still a few things missing:

- [x] Update `ast_iterator`
- [x] Fix remaining failing tests

However since this change will require coordinated action from the `ppxlib` people (cc @pitag-ha) and will impact other projects relying on parsetree / typedtree manipulation, I think it is useful to have the PR opened even if it is not completely polished.

Most of the changes are already implemented, namely updating the representation of `Longident.t`:

```ocaml
type t =
    Lident of string Location.loc
  | Ldot of t * string Location.loc
  | Lapply of t * t
```

I faced a number of breakage in the testsuite that, at least for some of them, are due to the use of polymorphic equality between longidents in the typer. It is very hard to identify the instances of these... I started fixing a bunch of these, but some tests are still failing, probably similar reason. Someone familiar with the (sub)-typing of first-class modules packages might be more lucky than me... 

Additionally, there is at least one instance of the use of the polymorphic `<=` between two longidents. I wrote an ad-hoc function to replace it, but I am unsure of the expected semantic. It looks like @garrigue introduced it in https://github.com/ocaml/ocaml/commit/d59bd570d72d4a0c7a78a97e6a451a7851b255fe. Could you have a look at this commit: aabf0225a4ee8b9fedac75e43e9be2170e269efe ? 

It would be extremely useful for this kind of refactoring to have an annotation that would tell the compiler not to accept polymorphic comparison of some types (which I think is a bit more precise than https://github.com/ocaml/ocaml/pull/9928).





For reference, here is the relevant part of the CI log for the first-class module test that is failing:

```diff
 Comparing program output firstclass.ml.corrected.corrected to reference firstclass.ml
 Action 3/3 (check-program-output) => failed (program output firstclass.ml.corrected.corrected differs from reference firstclass.ml: 
 --- firstclass.ml	2024-07-15 17:20:04.000000000 +0000
 +++ firstclass.ml.corrected.corrected	2024-07-15 17:32:01.230859744 +0000
 @@ -28,17 +28,12 @@
  module type S2 = sig type u type t type w end
  val g2 : (module S2 with type t = int and type u = bool) -> (module S') =
    <fun>
 -val h : (module S2 with type t = 'a) -> (module S with type t = 'a) = <fun>
 +Line 3, characters 10-75:
 +3 | let h x = (x : (module S2 with type t = 'a) :> (module S with type t = 'a));;
 +              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 +Error: Type "(module S2 with type t = 'a)" is not a subtype of
 +         "(module S with type t = 'a)"
 +       Modules do not match: S is not included in S2
         The type "w" is required but not provided
  |}];;
```

voodoos committed:
Promote changes in test: errors have finer lid locations

Also fix tests that rely on constructing longidents.

voodoos committed:
Remove polymorphic comparisons of longidents

The added locations in longidents change the behaviour of the generic
comparison functions.

Reduce diff in typecore

garrigue commented:
I suppose that you are talking of `n >= n2` at line 2217.
Here the order does not matter as long as it is total, so I suppose that your code is correct.
However, it feels wrong to include locations inside `type_expr` other than for debugging reasons.
So I would rather suggest switching to a different representation here, probably just a `string list` would do the job here, as one can only specify constraints for types in submodules.

voodoos commented:
> So I would rather suggest switching to a different representation here, probably just a `string list` would do the job here, as one can only specify constraints for types in submodules.

Thanks for the prompt feedback! I did felt like having opened Pandora's box when the typing errors started pilling up unexpectedly...

Just to clarify, looking at the `Types` AST there is only one occurrence of a `Longident.t`:
```ocaml
and type_desc =
  | Tpackage of Path.t * (Longident.t * type_expr) list
```

This is the one you suggest replacing by a `string list` (which would be the result of flattening the corresponding longident)?

voodoos committed:
Stop using longidents in Types.Tpackage

voodoos commented:
@garrigue (and cc @gasche since we discussed it earlier today): I tried replacing the `Longident.t` by a `string list` and all the typing tests are now passing. It is probably worth double-checking 🙂 (The main question being: are these longidents always "flat" ?)

There is still one failing test, but it is much less worrying, (ast diffing fails without any actual diff). I think the PR is ready for review.

gasche commented:
cc @Julow this is non-urgent, but I wonder if you would be interested in reviewing this

pitag-ha commented:
To quickly come back to

> However since this change will require coordinated action from the ppxlib people (cc @pitag-ha)

Let me forward the ping to @NathanReb and @patricoferris about this. As a quick note: Hopefully, this won't break too many PPXs. One pain point might be that a few PPXs might break semantically (while staying intact syntactically) due to polymorphic compares of longidents (i.e. what @voodoos ran into in this PR). That would be harder to catch than the typical syntactic PPX breakages we usually deal with.

In any case, thanks for this change, @voodoos! That'll make life easier for multiple tools/users.

Octachron commented on testsuite/tests/parsetree/test.reference:
```diff
@@ -0,0 +1,2 @@
+source.ml:  FAIL, REPARSED AST IS DIFFERENT
```
This test failure should probably not be promoted.

Julow replied:
I fixed the failure in this PR: https://github.com/voodoos/ocaml/pull/2

voodoos replied:
Nice catch, thanks a lot !

Julow reviewed (commented):
I have carefully reviewed this and found no obvious problems in the code. Some notes:

- Missing `location` calls in `Ast_iterator`, `Tast_iterator` and `Tast_mapper`.
- Lidents could possibly be passed to the polymorphic compare, which can now return a different result. I couldn't find a Map or a Set on long idents but I've done a manual search and there could still be one.
  We are hoping that the typed grep tool that @nojb showcased in https://discuss.ocaml.org/t/ann-2nd-editor-tooling-dev-meeting-25th-of-july/14953 will help find uses of `=` and `compare` on Longidents.
  This change will be very beneficial to OCamlformat and easy to handle in Odoc but might be problematic in projects that depend on the unstable compiler-libs and the incompatible type might not help catch this issue. We must make the community aware of this change.

gasche commented:
> Lidents could possibly be passed to the polymorphic compare, which can now return a different result. I couldn't find a Map or a Set on long idents but I've done a manual search and there could still be one.

As I suggested privately to @voodoos, a solution to detect this is to temporarily change the `Longident.t` definition to add a function closure as an extra first parameter of each constructor. Fix the code to build, run the testsuite, any polymorphic comparison will now give you a backtrace.

Julow commented:
@gasche I have tried that for a bit (changing the `loc` type for easier refactoring) but this also interferes with marshalling and prevents the compiler from building the stdlib. I went not very far, with this branch: https://github.com/Julow/ocaml/tree/13302_location_non_comparable The first commit removes most places where the `loc` type is constructed, could that be useful on its own?

voodoos commented:
@gasche @Julow we found one more instance  with @Octachron in [3e3ef8f](https://github.com/ocaml/ocaml/pull/13302/commits/3e3ef8fc75f4b7bc7b88db0a1dfda01f56e67eec) following an alternative approach: make compiling fail in translprim / typeopt when some optimization are done on comparison. 

It only works when the type is readily accessible however and @Octachron suggested other exotic solutions like adding a large number of constructor to the longident type or a specific integer and fail when polymorphic comparison hits these but we didn't try it.

voodoos commented:
>     * Missing `location` calls in `Ast_iterator`, `Tast_iterator` and `Tast_mapper`.

I pushed a commit that update these iterators.

Julow commented:
> It only works when the type is readily accessible however and @Octachron suggested other exotic solutions like adding a large number of constructor to the longident type or a specific integer and fail when polymorphic comparison hits these but we didn't try it.

I tried this with polymorphic variants [in this branch](https://github.com/Julow/ocaml/tree/13302_detect_lident_compare). It only finds the case you fixed last week and the safe one in [the parsetree test](https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/parsetree/test.ml#L72).
Perhaps this should be tried on more code ?

gasche commented:
I think that the amount of testing on this question (changes in polymorphic comparison behavior) is now adequate, we can consider that the current behavior is correct. This needs a review for correctness to be mergeable.

voodoos committed:
Attach new locs to `Ldot` and `Lapply` nodes
instead of `Lident` nodes.

Parser changes.
Requires bootstrap.

Co-authored-by: Jules Aguillon <jules@j3s.fr>
Suggested-by: Florian Angeletti <florian.angeletti@inria.fr>

voodoos committed:
Update dependency info

voodoos committed:
Add changelog entry for #13302

voodoos commented:
I rebased and cleaned the history a bit.

@Octachron the hygiene checks complains that there is no parser changes so we should remove the label. Does it checks that the recognized grammars are the same or that the resulting parsetree is the same ?

Octachron commented:
The `parsetree_change` looks for changes in the parsetree. The issue here is that the change is in a dependency of the parsetree while the parsetree itself is untouched. Extending the check to include `Longident` is probably a good idea too.

voodoos commented:
But ideally we should only check that `Longident.t` changed since there are a few utility functions defined this module, it might be tricky to do well.

voodoos commented:
> This needs a review for correctness to be mergeable.

@gasche do you think you could be that reviewer ? I pinged around with not a lot of success...

NathanReb commented:
The parestree changes issue should be fixed by https://github.com/ocaml/ocaml/pull/13431. I think that even though not all API changes to those modules will be breaking, it's good for ppxlib maintainers to keep track of them.

voodoos commented:
> The parestree changes issue should be fixed by #13431. I think that even though not all API changes to those modules will be breaking, it's good for ppxlib maintainers to keep track of them.

Thank you @NathanReb ! I rebased so that we can see if the CI passes now 🙂 (spoiler, it looks like it doesn't... oh right https://github.com/ocaml/ocaml/pull/13431 is not merged yet, my bad).

That being said, you might be interested in reviewing these changes since they will certainly have an impact on the PPX ecosystem. Would you happen to have some spare cycles for this ?

voodoos commented:
> > The parestree changes issue should be fixed by #13431. I think that even though not all API changes to those modules will be breaking, it's good for ppxlib maintainers to keep track of them.
> 
> Thank you @NathanReb ! I rebased so that we can see if the CI passes now 🙂 (spoiler, it looks like it doesn't oh right #13431 is not merged yet, my bad).
> 
> That being said, you might be interested in reviewing these changes since they will certainly have an impact on the PPX ecosystem. Would you happen to have some spare cycles for this ?

NathanReb commented:
Sure, I'll take a look!

NathanReb commented:
From a quick look I think only the change in the `Longident.t` type will impact us. It hadn't changed until now so it is not versioned by ppxlib but that should not be too hard to deal with. This is good to go form my end!

voodoos commented:
I rebased the PR after our (offline) discussion, @gasche . Do you think that it needs another independent review ? Could it be yours ?

Octachron commented on parsing/longident.mli:
```diff
@@ -23,11 +23,17 @@
-  | Ldot of t * string
+    Lident of string loc
+  | Ldot of t * string loc
   | Lapply of t * t
```
Experimenting with b21244a7b13d2c3e3f74136edd647e169b3fa7da, my impression is that it would be less intrusive to store locations when we split the parent location and avoid storing redundant locations on `Lident`:
```ocaml
type t =
| Lident of string
| Ldot of t loc * string loc
| Lapply of t loc * t loc
   (* I have the impression that those Lapply locations are simply missing in the present PR? *)
``` 
This has the advantage of avoiding the quite frequent redundant form `mkloc (Lident (mkloc "ident" loc) loc` at the cost of introducing more locations for long `Longident`. 

Another potential option might to have an explicit notion of location splitting and defining `Longident.t` as

```ocaml
type t =
| Lident of string
| Ldot of t * string * Loc.split
| Lapply of t loc * t loc 
```

Similarly, I have the feeling that it might be good idea to define constructors functions for the `Ldot` and `Lapply` case:
```ocaml
let dot ?lloc p ?rloc s = ...
let app ?floc f ?xloc x = ...
```

voodoos replied:
Yes, it seems worth it to remove the redundants `mkloc (Lident (mkloc "ident" loc) loc`. 
Do you have a feeling about which of your two propositions would be the best ?

If the additional locs of the `| Ldot of t loc * string loc` version are not concerning in term of space, it looks like the most useful for tooling and does not introduce additional complexity to interpret the locations as does the `Loc.split` version.

Octachron replied:
I forgot to include my analysis of the number of stored locations: if we have a `Longident.t loc` with `n` components, we store:

1. one location before this PR
2.  n + 1 locations with this PR
3. 2 * (n - 1) + 1 with my first proposition
4. (n-1) splits + 1 location with the alternate one

In particular, for single component `Longident.t`, the representations 1., 3., 4. are the more efficient.
Representation 3 is the worse for long `Longident.t` however.

And on average, the representation 3 will be lighter than representation 2 if the average length of Longident.t is less than 2.

That sounds like a reasonable hypothesis to me, but I will try to check it empirically soon.

voodoos replied:
The question is: is it worth the trouble compared to the initial naive solution ?

Octachron replied:
Computing some longident statistics with https://github.com/Octachron/ocaml-idioms-statistics, the average `Longident.t` length seems to varies between `1.03` (seen for `graphics`) and `1.38` (seen for `ppxlib`) depending on projects (with a total average around `1.15`?). Thus my first proposal is probably good enough.

voodoos replied:
I went for that option in [dc603a0](https://github.com/ocaml/ocaml/pull/13302/commits/dc603a0de2cc2640ef6efd7694022411854244cc). As expected, the diff is much smaller, and it revealed a natural opportunity to improve the locations in error messages that concern only a prefix of a longident or a functor's argument.

Octachron commented on parsing/printast.ml:
```diff
@@ -37,8 +37,9 @@ let fmt_location f loc =
-  | Longident.Ldot (y, s) -> fprintf f "%a.%s" fmt_longident_aux y s
+  | Longident.Lident { txt = s; _ } -> fprintf f "%s" s
+  | Longident.Ldot (y, { txt = s; _ }) ->
+      fprintf f "%a.%s" fmt_longident_aux y s
```
For the ast printers, we probably want to print the new location information.

voodoos replied:
I started thinking about it and I am wondering if it's really a good thing to do. There is a lot of new locations and it might clutter the output a lot.

Octachron commented on typing/includemod.ml:
```diff
@@ -229,7 +229,11 @@ module Core_inclusion = struct
+        Ok (Includecore.value_descriptions ~loc env (Ident.name id) vd1 vd2)
+      in
+      (* Cmt_format.record_uid_dependency vd1.val_uid vd2.val_uid; *)
+      check
```
This part looks like a rebasing remnant.

Octachron commented on typing/out_type.ml:
```diff
@@ -589,7 +590,7 @@ let is_unambiguous path env =
       (* also allow repeatedly defining and opening (for toplevel) *)
       let id = lid_of_path p in
-      List.for_all (fun p -> lid_of_path p = id) rem &&
+      List.for_all (fun p -> Longident.same (lid_of_path p) (*arg*) id) rem &&
```
I am not sure what the comment `(* arg *)` means here?

Octachron commented on typing/typecore.ml:
```diff
@@ -1400,7 +1403,9 @@ let disambiguate_label_by_ids closed ids labels  : (_, _) result =
-  let ids = List.map (fun (lid, _) -> Longident.last lid.txt) lid_a_list in
+  let ids =
+    List.map (fun (lid, _) -> Longident.last lid.txt) lid_a_list
+  in
```
This looks like some refactoring noise.

Octachron commented on typing/typecore.ml:
```diff
@@ -5529,8 +5538,8 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
-         exp_desc =
-         Texp_ident(Path.Pident id, mknoloc (Longident.Lident name), desc)}
+         exp_desc = Texp_ident(Path.Pident id,
+                               mknoloc (Longident.Lident name), desc)}
```
This is now an unrelated style change.

Octachron commented on typing/typecore.ml:
```diff
@@ -1228,7 +1230,8 @@ end) = struct
       warn lid.loc
-        (Warnings.Name_out_of_scope (path_s, [Longident.last lid.txt], false))
+        (Warnings.Name_out_of_scope
+          (path_s, [Longident.last lid.txt], false))
```
This is a style change now.

Octachron commented on testsuite/tools/expect.ml:
```diff
@@ -78,7 +78,7 @@ let match_expect_extension (ext : Parsetree.extension) =
               [ a
               ; { pexp_desc = Pexp_construct
-                                ({ txt = Lident "Principal"; _ }, Some b) }
+                  ({ txt = Lident "Principal"; _ }, Some b) }
```
This is refactoring ghost.

Octachron commented on testsuite/tests/compiler-libs/test_longident.ml:
```diff
@@ -6,20 +6,24 @@
 
 module L = Longident
+let mknoloc = Location.mknoloc
+let lident l = mknoloc (L.Lident l)
```
This test is bit too noisy with all locations, I propose to use a pretty-printer for locations:
```diff
--- a/testsuite/tests/compiler-libs/test_longident.ml
+++ b/testsuite/tests/compiler-libs/test_longident.ml
@@ -1,12 +1,15 @@
 (* TEST
- flags = "-I ${ocamlsrcdir}/parsing";
- include ocamlcommon;
+ flags = "-I ${ocamlsrcdir}/parsing -I ${ocamlsrcdir}/toplevel";
+ include ocamltoplevel;
  expect;
 *)
 [@@@alert "-deprecated"]
 
 module L = Longident
 let mknoloc = Location.mknoloc
+let () =
+  Topdirs.dir_install_printer Format.err_formatter
+   (L.parse "Location.print_loc")
```

voodoos replied:
Done in [ef7882d](https://github.com/ocaml/ocaml/pull/13302/commits/ef7882d19cbf80fac1d6473181f7962bf41d4100)

Octachron commented on debugger/debugger_parser.mly:
```diff
@@ -188,7 +189,8 @@ longident_eol :
   | LIDENT                      { Some (Lident $1) }
-  | module_path DOT UIDENT      { Some (Ldot($1, $3)) }
+  | module_path DOT UIDENT
+      { Some (Ldot(Location.mknoloc $1, Location.mknoloc $3)) }
```
The parser style is more to keep the left `{`s aligned when possible.

voodoos replied:
Done in [b77adc6](https://github.com/ocaml/ocaml/pull/13302/commits/b77adc6fd2d44fc7b16c9b10a5516ef02bf811a8) ?

Octachron reviewed (approved):
I find that this version strikes a better balances between intrusiveness and new features. The improvement to the error messages are a nice bonus.

Moreover, it feels like this version would be easier to revisit if we wish to switch to a more compact representation. Overall, I agree that we really want to avoid developer tools reparsing `Longident` (because this is error prone).

Nevertheless, I have few remarks left that you may want to fix before merging.

voodoos commented:
Thank you for the review, I will squash the two implementation commits together before re-basing on trunk and address your comments.

voodoos committed:
bootstrap

voodoos committed:
Less verbose longident test output

Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>

voodoos commented:
@Octachron I rebased on trunk once more and addressed your comments in the last commits.
I don't know if I should do something about the hygiene of the debugger parser ?

Octachron commented on debugger/debugger_parser.mly:
```diff
@@ -171,15 +171,16 @@ opt_signed_int64_eol :
-  | module_path DOT LPAREN OPERATOR RPAREN { Ldot($1, $4) }
+  | module_path DOT OPERATOR    { Ldot(Location.mknoloc $1, Location.mknoloc $3) }
+  | module_path DOT LPAREN OPERATOR RPAREN
+                                { Ldot(Location.mknoloc $1, Location.mknoloc $4) }
```
Since nearly all rules are too long, I propose to add a newline after the production, and align from there
```ocaml
    LIDENT
      { Lident $1 }
  | module_path DOT LIDENT
      { Ldot(Location.mknoloc $1, Location.mknoloc $3) }
  | OPERATOR
      { Lident $1 }
  | module_path DOT OPERATOR
      { Ldot(Location.mknoloc $1, Location.mknoloc $3) }
  | module_path DOT LPAREN OPERATOR RPAREN
      { Ldot(Location.mknoloc $1, Location.mknoloc $4) }
 ```

Octachron commented:
@voodoos , what is the current base of the PR? The failure in the CI looks like #13090 which should have been fixed in the latest trunk.

voodoos commented:
> @voodoos , what is the current base of the PR? The failure in the CI looks like #13090 which should have been fixed in the latest trunk.

It is based on yesterday's dbb4a318b7824a1e1f4ba5223a5cf08d8ba7340c

Octachron commented on parsing/ast_iterator.ml:
```diff
@@ -89,6 +89,20 @@ let iter_opt f = function None -> () | Some x -> f x
+
+let iter_loc_lid sub {loc; txt} =
+  iter_loc_lid sub txt;
+  iter_loc sub {loc; txt}
```
Sorry, I missed that point earlier, but we probably want to have `iter_loc` first to keep the iterator a depth-first traversal.

Octachron commented on typing/tast_iterator.ml:
```diff
@@ -68,6 +68,20 @@ type iterator =
+
+let iter_loc_lid sub {loc; txt} =
+  iter_loc_lid sub txt;
+  iter_loc sub {loc; txt}
```
Same remark as on the parsetree iterator: iterating first on the full location before iterating on the sublocation feels slightly more consistent to me.

voodoos committed:
Ensure depth-first search in iterators

Octachron reviewed (approved):
I agree that avoiding re-parsing and heuristic on `Longident.t` is a good reason to store the components of subcomponents of the `Longident`, and the improved error messages are nice.
